### ISSUE TITLE
„Rustification“ of the API

### ISSUE THREAD

**Issue by vorner** [fede1024/rust-rdkafka#73]
Hello

When using the library, I have somewhat non-rusty feeling from the API. To highlight some (this is not exhaustive):

* The `ClientConfig::set` that takes two string arguments and panics at runtime if it is wrong could be done with an enum, eg. `.set(ConfigValue::BootstrapServers(&["srv1", "srv2"]))` ‒ which would ensure that the option is valid and has the right type at compile time.
* The consumer could provide an iterator API to just read the messages, without caring too much about what timed out, etc.
* The numeric parameter to `consumer.poll` is error-prone. The better way would be to use a `Option<Duration>` as the parameter (with `None` meaning no timeout).

I understand this is what the wrapped library provides and that it would take extra work to provide rust-native API. What I want to ask is if there's another reason for it.

In other words, if I started sending PRs that change the API to look more rust-like, would you be willing to accept them (provided they satisfy needed quality requirements), or would it be against the goal of the library?

**Comment by fede1024** [comment]
I would definitely be in favor! :) I think we should keep the string based configuration API for tools that want to inject it from configuration files, but it definitely doesn't mean that it should be the only one available. In particular having Rust enums with the configuration options (as in your example) would make configuration discovery much easier, thanks to autocompletion etc.

So yeah, if you could contribute in that direction it would be awesome :) (same goes for your other proposals).

**Comment by thijsc** [comment]
👍 on keeping the string based api, we use that to load the config from file.

**Comment by thijsc** [comment]
It'd be nice to have typed api for the config. It does incur a maintenance burden somebody will have to pay though, we have to check if anything changed in the config everytime the librdkafka release is bumped.

**Comment by benesch** [comment]
> The consumer could provide an iterator API to just read the messages, without caring too much about what timed out, etc.

This exists now!

> The numeric parameter to consumer.poll is error-prone. The better way would be to use a Option<Duration> as the parameter (with None meaning no timeout).

As does this! APIs that take a timeout accept anything that can be turned into a new [`Timeout`](https://docs.rs/rdkafka/0.27.0/rdkafka/util/enum.Timeout.html) type.

The enum-based configuration API would be really nice, but would require more maintenance bandwidth than we currently have available. So I'm going to close this out, but if someone wants to contribute that enum-based API, they should go for it!
